### PR TITLE
build: run tests on node.js v21 nightly

### PR DESCRIPTION
**Pull request by aladdin-add**

### CODE REVIEW

**Comment by iambumblehead** [comment]
Hey everyone I tried resolving things here but encountered the upstream error reported here https://github.com/nodejs/node/issues/47614

Also, there seems to be a separate issue where the loader isn't engaged by the ava unit tests. I tried using this alternative syntax in the tests-ava package.json but the loader wasn't engaged...
```json
{
  "ava": {
    "nodeArguments": [
      "--loader=esmock"
    ]
  }
}
``` 
Feel free to leave any input or advice for me. Thanks for pushing v20 tests @aladdin-add

**Comment by iambumblehead** [comment]
the ava issue is reproduced and reported here https://github.com/avajs/ava/issues/3195

**Comment by aladdin-add** [comment]
as the `--loader` is still experimental, I like to add a task to test it on node.js nightly build, and file an issue if it fails. WDYT?

will give it a try if no objections. 😄

**Comment by aladdin-add** [comment]
looks like the node.js v20-nightly is still failing.

**Comment by iambumblehead** [comment]
this branch is rebased on the latest master which handles async `import.meta.resolve` differently and specifies node-versions <=19.9.0

**Comment by iambumblehead** [comment]
The nightly [used in the tests](https://github.com/iambumblehead/esmock/actions/runs/4928686213/jobs/8899211986?pr=199#step:3:8) appears to be an older nightly 20.0.0-nightly202304181c3d741cb0 as indicated by the timestamp in the name. We will want 20230514 seen here https://nodejs.org/download/nightly/

**Comment by iambumblehead** [comment]
tests for all folders fail when run from the latest nightly... I won't have bandwidth to investigate further until end of week

**Comment by iambumblehead** [comment]
if anyone here knows generally the way we should data should be communicated from test to loader, feel free to leave incomplete advice. If we there is a generally recommended approach, I might be able to add this this evening

**Comment by iambumblehead** [comment]
let's try to implement this way, if there are no other suggestions https://github.com/nodejs/node/pull/44710#issuecomment-1368192020

**Comment by iambumblehead** [comment]
I have a passing test here that uses worker thread to define the longer uri in the loader, but cannot use this machine to push up any changes so it might take some time before anything reaches this branch

**Comment by iambumblehead** [comment]
I thought these changes were developed around node 21 but it looks like node 19 was used :/

**Comment by iambumblehead** [comment]
When using node 21 nightly here, changes fail for ava's test folder, but they pass for the other test-runners.

File: src/esmockCache.js [hunk 1]
```diff
@@ -7,6 +7,8 @@ const esmockCache = {
 }
 
 const esmockTreeIdSet = (key, keylong) => (
+  (typeof global.postMessageEsmk === 'function')
```
Comment (koshic) [inline src/esmockCache.js:10]: 'old' and 'new' variants are mutual exclusion - the old one doesn't work with dedicated thread, and the new one doesn't exist if globalPreload hook is not supported.

I propose to split those branches explicitly and choose the way based on postMessageEsmk existance.

**Review by koshic — CHANGES_REQUESTED** [review]

**Comment by iambumblehead** [comment]
ava does not work with node v20+ loaders at all https://github.com/iambumblehead/nodejs-v20-loader-ava-noop

File: src/esmockCache.js (hunk 1 above)
Comment (iambumblehead) [reply src/esmockCache.js:10]: its _almost_ working we just need to disable test-tsm, which fails with this message
```bash
error: Invalid version: "21.0.0-nightly202305158b3777d0c8"
```

Comment (iambumblehead) [reply src/esmockCache.js]: It seems a split is not needed, but perhaps we should try using an unary here or some other mechanism for reducing the number of definitions created.  I would prefer to make such changes in a follow up PR just to this PR along and publish it so esmock will work with the latest node versions again

Comment (iambumblehead) [reply src/esmockCache.js]: I'll try pushing up changes with an unary to resolve this discussion

Comment (iambumblehead) [reply src/esmockCache.js]: this change caused the tests to fail https://github.com/iambumblehead/esmock/actions/runs/4990343286/jobs/8935419674?pr=199 I have an idea around why it happens but am sleepy its 2:45am here (reverting)
```diff
diff --git a/src/esmockCache.js b/src/esmockCache.js
index 51c625b..0420787 100644
--- a/src/esmockCache.js
+++ b/src/esmockCache.js
@@ -8,8 +8,8 @@ const esmockCache = {
 
 const esmockTreeIdSet = (key, keylong) => (
   typeof global.postMessageEsmk === 'function'
-    && global.postMessageEsmk({ key, keylong }),
-  global.mockKeys[String(key)] = keylong)
+    ? global.postMessageEsmk({ key, keylong })
+    : global.mockKeys[String(key)] = keylong)
 
 const esmockTreeIdGet = key => (
   global.mockKeys[String(key)])

```

Comment (iambumblehead) [reply src/esmockCache.js]: Politely, I'd like to mark this as resolved. The change was requested when tests were still failing but now they pass (when the diff from the previous comment is not applied).

**Comment by flovogt** [comment]
title says enable node v20 but currently only nightly-v21 is used. That should be reflected in the title or in the code

**Comment by iambumblehead** [comment]
@flovogt okay, which do you recommend?

**Comment by iambumblehead** [comment]
@flovogt the title is updated

**Comment by koshic** [comment]
I think it's better to merge it ASAP and publish new package version. We can improve loader-userspace logic interaction later.

**Review by koshic — APPROVED** [review]

**Comment by iambumblehead** [comment]
a new PR adding node v20 tests to the ci pipeline is created https://github.com/iambumblehead/esmock/pull/202

**Pull request merged**